0049-protocol-versioning.md

```diff
@@ -0,0 +1,42 @@
+
+Besides this, we also want a way to programatically reason about different versions of the protocol from within our software, including having the ability to be aware of protocol versions within the snark proofs that Mina uses.
+
+The solution in this RFC is optimizing for simplicity and clarity.
+
+## Detailed design
+
+At a protocol level, there are 2 important dimensions of compatibility: compatibility of the transaction system, and compatibility of the networking protocol. Transation system compatibility determines support for the transaction format and the logical details of how transactions are processed. Compatibility of the networking protocol determines the participation rules (consensus et al) and wire format.
+
+Compatibility of the transaction system can be thought of as the mapping of ledger states into sets of all valid applicable transactions for each ledger state. This insures that, if we have any 2 different implementations of the transaction logic, that those 2 implementations are only considered compatible if they will always accept, reject, or fail transactions in an equivalent way.
+
+Compatibility of the networking protocol can be thought of as the set of all RPC messages, wires types, gossip topics, p2p feature set, and participation rules. Participation rules include any rule in the protocol that regards certain behavior as malicious or otherwise not permitted. This ranges from consensus details to bannable offenses and even verification logic for gossip messages. It's important the capture all of this under the umbrella of networking protocol compatibility since divergences in these details can lead to unintended forks on the chain.
+
+To label versions of our daemon software, we will use the following versioning convention, inspired by semver: `<txn_version>.<net_version>.<impl_version>`. In this setup, the version of the transaction system is the most dominate version, since any updates to it necessitate a hardfork, and when reasoning about the logic of the chain state, it is the most significant version number. The prefix of the `<txn_version>.<net_version>` uniquely identifies a particular hard fork version, since the pair of those 2 versions determines the full compatability of an implementation. The leftover `<impl_version>` is retained for the usage of individual daemon implementations to use however they see fit. We leave this detail to each implementor of the daemon, as the meaning of these versions is intended to be specific to the implementation.
```

Comment:
Would it make sense to label the components as distinct values?

Writing something like `3.2` suggests there was a sequence of versions 3.0, 3.1, 3.2.

On the other hand, writing 
```
transaction_version = 3
networking_version = 2
```
communicates the independence of those components.

Reply:
Well, I was imagining that it would be a sequence still. Bumping the transaction version necessitates a new hard fork version, so we reset the networking version back to 0. The networking version is "networking version within this transaction version". In other words, if we were at `3.2`, and then update the transaction version, we would go to `4.0` even if the networking implementation didn't change, since this is the first networking version supported that supports this transaction version. We would always thus think about compatibility of the daemon as the combination of these 2 versions. The idea here is that whether or not the changes effect transactions or networking informs us which number we should be incrementing.